File: backend/app/audit.py
"""Append-only, auto-redacting audit log persisted to Postgres/SQLite.

Every security-critical action is recorded with references and non-sensitive
metadata only. A redaction pass strips anything that looks like a credential
before the row is written, so secrets never reach the DB via the audit path.
"""
from __future__ import annotations

import re

from sqlalchemy import select

from .models import AuditEvent

_SENSITIVE = re.compile(r"pass|pwd|secret|token|cookie|card|cvc|cvv|otp|3ds|ssn|captcha", re.I)


def redact(value, depth: int = 0):
    if value is None or depth > 5:
        return value
    if isinstance(value, str):
        return value[:32] + "…[redacted]" if len(value) > 200 else value
    if isinstance(value, list):
        return [redact(v, depth + 1) for v in value]
    if isinstance(value, dict):
        out = {}
        for k, v in value.items():
            out[k] = "[redacted]" if _SENSITIVE.search(str(k)) else redact(v, depth + 1)
        return out
    return value


def record(db, *, org_id: str, application_id: str, action: str, detail=None,
           actor: str = "system") -> AuditEvent:
    from sqlalchemy import func
    next_seq = (db.query(func.max(AuditEvent.seq)).scalar() or 0) + 1
    ev = AuditEvent(seq=next_seq, org_id=org_id, application_id=application_id, actor=actor,
                    action=action, detail=redact(detail or {}))
    db.add(ev)
    db.commit()
    db.refresh(ev)
    return ev


def for_application(db, application_id: str):
    return db.execute(
        select(AuditEvent).where(AuditEvent.application_id == application_id).order_by(AuditEvent.seq)
    ).scalars().all()


def contains_plaintext(db, needle: str) -> bool:
    rows = db.execute(select(AuditEvent)).scalars().all()
    return any(needle in str(r.detail) for r in rows)

File: src/main/portal/audit.mjs
// Append-only audit log for security-critical actions. Entries are immutable
// once written; secrets are never recorded — only references, non-sensitive
// metadata, and outcomes. A redaction pass strips anything that looks like a
// credential before an entry is stored.

const SENSITIVE_KEYS = /pass|pwd|secret|token|cookie|card|cvc|cvv|otp|3ds|ssn/i

function redact(value, depth = 0) {
  if (value == null || depth > 4) return value
  if (typeof value === 'string') {
    // References like vault://... and RCPT-... are safe; long opaque blobs get masked.
    return value.length > 200 ? value.slice(0, 32) + '…[redacted]' : value
  }
  if (Array.isArray(value)) return value.map((v) => redact(v, depth + 1))
  if (typeof value === 'object') {
    const out = {}
    for (const [k, v] of Object.entries(value)) {
      out[k] = SENSITIVE_KEYS.test(k) ? '[redacted]' : redact(v, depth + 1)
    }
    return out
  }
  return value
}

export class AuditLog {
  constructor(clock = () => Date.now()) {
    this._entries = []
    this.clock = clock
  }

  /** Append an immutable, redacted audit event. */
  record(caseId, action, detail = {}, actor = 'system') {
    const entry = Object.freeze({
      seq: this._entries.length + 1,
      at: this.clock(),
      caseId,
      actor,
      action,
      detail: redact(detail)
    })
    this._entries.push(entry)
    return entry
  }

  forCase(caseId) { return this._entries.filter((e) => e.caseId === caseId) }
  all() { return [...this._entries] }
  /** Detect any accidental secret leakage in the whole log. */
  containsPlaintext(needle) { return JSON.stringify(this._entries).includes(needle) }
}
